File: ruby/TEST_07_socket.rb
#!/usr/bin/env ruby
# coding: utf-8
# vim: set sw=2 sts=2 et nowrap fenc=utf-8 :
# Copyright 2010 Ali Polatel <alip@exherbo.org>
# Distributed under the terms of the GNU General Public License v2

%w{socket test/unit}.each {|m| require m}

$: << File.expand_path('.libs')
require 'PinkTrace'

class TestPinkSocket < Test::Unit::TestCase
  def test_socket_name_invalid
    assert_raise ArgumentError do
      PinkTrace::Socket.name
    end
    assert_raise ArgumentError do
      PinkTrace::Socket.name 1, 2
    end
    assert_raise TypeError do
      PinkTrace::Socket.name 'pink'
    end
  end

  def test_socket_decode_call_invalid
    assert_raise ArgumentError do
      PinkTrace::Socket.decode_call
    end
    assert_raise ArgumentError do
      PinkTrace::Socket.decode_call 0, 1, 2
    end
    assert_raise TypeError do
      PinkTrace::Socket.decode_call 'pink'
    end
    assert_raise TypeError do
      PinkTrace::Socket.decode_call 0, 'pink'
    end
  end

  def test_socket_decode_call_esrch
    assert_raise Errno::ESRCH do
      PinkTrace::Socket.decode_call 0
    end
  end

  def test_socket_decode_fd_invalid
    assert_raise ArgumentError do
      PinkTrace::Socket.decode_fd
    end
    assert_raise ArgumentError do
      PinkTrace::Socket.decode_fd 0, 1, 2, 3
    end
    assert_raise TypeError do
      PinkTrace::Socket.decode_fd 'pink'
    end
    assert_raise TypeError do
      PinkTrace::Socket.decode_fd 0, 'pink'
    end
    assert_raise TypeError do
      PinkTrace::Socket.decode_fd 0, 1, 'pink'
    end
    assert_raise PinkTrace::BitnessError do
      PinkTrace::Socket.decode_fd 0, 1, 13
    end
    assert_raise PinkTrace::IndexError do
      PinkTrace::Socket.decode_fd 0, PinkTrace::MAX_INDEX
    end
  end

  def test_socket_decode_fd_esrch
    assert_raise Errno::ESRCH do
      PinkTrace::Socket.decode_fd 0
    end
  end
end

# These test cases depend on generated system call names.
# Don't run them if they weren't generated.
unless PinkTrace::SysCall.name 0
  exit 0
end

class TestPinkSocket
  TEST_SOCKET = './TEST_unix_socket'

  def teardown
    begin
      File.unlink TEST_SOCKET
    rescue Errno::ENOENT
    end
  end

  def test_decode_socket_call
    pid = PinkTrace::Fork.fork do
      s = UNIXServer.new './TEST_unix_socket'
    end

    # Loop until we get to the socket() system call as there's no guarantee
    # that other system calls won't be called beforehand.
    event = -1
    while event != PinkTrace::Event::EXIT_GENUINE
      PinkTrace::Trace.syscall pid
      Process.waitpid pid

      event = PinkTrace::Event.decide
      if event == PinkTrace::Event::SYSCALL then
        scno = PinkTrace::SysCall.get_no pid
        name = PinkTrace::SysCall.name scno
        if name == 'socketcall' or name == 'socket' then
          subcall = PinkTrace::Socket.decode_call pid
          if name == 'socketcall' then
            # The call must have given the decoded socket call.
            subname = PinkTrace::Socket.name subcall
          else
            subname = PinkTrace::SysCall.name subcall
          end
          assert(subname == 'socket', "Wrong subcall name, expected: socket got: '#{subname}'")
          break
        end
      end
    end

    begin PinkTrace::Trace.kill pid
    rescue Errno::ESRCH ;end
  end

  def test_decode_socket_fd
  end
end

File: python/TEST_07_socket.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# vim: set sw=4 ts=4 sts=4 et :

import os, socket, sys, unittest

sys.path.insert(0, '.')
import pinktrace

class TestSocket_01_Invalid(unittest.TestCase):

    def test_01_name(self):
        self.assertRaises(TypeError, pinktrace.socket.name)

    def test_02_decode_call(self):
        self.assertRaises(TypeError, pinktrace.socket.decode_call)
        self.assertRaises(TypeError, pinktrace.socket.decode_call, 'pink')
        self.assertRaises(ValueError, pinktrace.socket.decode_call, 0, 13)

    def test_03_decode_fd(self):
        self.assertRaises(TypeError, pinktrace.socket.decode_fd)
        self.assertRaises(IndexError, pinktrace.socket.decode_fd, 0, pinktrace.syscall.MAX_INDEX)
        self.assertRaises(ValueError, pinktrace.socket.decode_fd, 0, 1, 13)

    def test_04_decode_address(self):
        self.assertRaises(TypeError, pinktrace.socket.decode_address)
        self.assertRaises(TypeError, pinktrace.socket.decode_address, 0)
        self.assertRaises(IndexError, pinktrace.socket.decode_address, 0, pinktrace.syscall.MAX_INDEX)
        self.assertRaises(ValueError, pinktrace.socket.decode_address, 0, 1, 13)

## FIXME: The test cases below must be skipped if system call names weren't generated.
TEST_UNIX_SOCKET = './TEST_UNIX_SOCKET'
class TestSocket_02(unittest.TestCase):

    def test_01_decode_call(self):
        pid = pinktrace.fork.fork()
        if pid == 0: # child
            s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
            os._exit(0)
        else: # parent
            # Loop until we get to the socket() system call as there's no
            # guarantee that other system calls won't be called beforehand.
            ev = -1
            while ev != pinktrace.event.EVENT_EXIT_GENUINE:
                pinktrace.trace.syscall(pid)
                pid, status = os.waitpid(pid, 0)

                ev = pinktrace.event.decide(status)
                if ev == pinktrace.event.EVENT_SYSCALL:
                    scno = pinktrace.syscall.get_no(pid)
                    name = pinktrace.syscall.name(scno)
                    if name == 'socketcall':
                        # Decode the call to its subcall
                        subno = pinktrace.socket.decode_call(pid)
                        subname = pinktrace.socket.name(subname)
                        self.assertEqual(subname, 'socket')
                        break
                    elif name == 'socket':
                        break

            try: pinktrace.trace.kill(pid)
            except OSError: pass

    def test_02_decode_fd(self):
        pid = pinktrace.fork.fork()
        if pid == 0: # child
            s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
            os._exit(0)
        else: # parent
            # Loop until we get to the socket() system call as there's no
            # guarantee that other system calls won't be called beforehand.
            ev = -1
            while ev != pinktrace.event.EVENT_EXIT_GENUINE:
                pinktrace.trace.syscall(pid)
                pid, status = os.waitpid(pid, 0)

                ev = pinktrace.event.decide(status)
                if ev == pinktrace.event.EVENT_SYSCALL:
                    scno = pinktrace.syscall.get_no(pid)
                    name = pinktrace.syscall.name(scno)
                    if name == 'socketcall' or name == 'socket':
                        fd = pinktrace.socket.decode_fd(pid)
                        self.assert_(fd > 0)

            try: pinktrace.trace.kill(pid)
            except OSError: pass

    def test_03_decode_address_unix(self):
        pid = pinktrace.fork.fork()
        if pid == 0: # child
            s = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
            s.connect(TEST_UNIX_SOCKET)
        else: # parent
            # Loop until we get to the connect() system call as there's no
            # guarantee that other system calls won't be called beforehand.
            ev = -1
            addr = None
            while ev != pinktrace.event.EVENT_EXIT_GENUINE:
                pinktrace.trace.syscall(pid)
                pid, status = os.waitpid(pid, 0)

                ev = pinktrace.event.decide(status)
                if ev == pinktrace.event.EVENT_SYSCALL:
                    scno = pinktrace.syscall.get_no(pid)
                    name = pinktrace.syscall.name(scno)
                    if name == 'socketcall':
                        subno = pintrace.socket.decode_call(pid)
                        subname = pinktrace.socket.name(subno)
                        if subname == 'connect':
                            addr = pinktrace.socket.decode_address(pid, 1)
                            break
                    elif name == 'connect':
                        addr = pinktrace.socket.decode_address(pid, 1)
                        break

            self.assert_(isinstance(addr, pinktrace.socket.Address), "%r" % addr)
            self.assertEqual(addr.family, socket.AF_UNIX)
            self.assertEqual(addr.path, TEST_UNIX_SOCKET)
            self.assertEqual(addr.ip, None)

            try: pinktrace.trace.kill(pid)
            except OSError: pass

    def test_04_decode_address_inet(self):
        pid = pinktrace.fork.fork()
        if pid == 0: # child
            s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
            s.connect(("127.0.0.1", 12345))
        else: # parent
            # Loop until we get to the connect() system call as there's no
            # guarantee that other system calls won't be called beforehand.
            ev = -1
            addr = None
            while ev != pinktrace.event.EVENT_EXIT_GENUINE:
                pinktrace.trace.syscall(pid)
                pid, status = os.waitpid(pid, 0)

                ev = pinktrace.event.decide(status)
                if ev == pinktrace.event.EVENT_SYSCALL:
                    scno = pinktrace.syscall.get_no(pid)
                    name = pinktrace.syscall.name(scno)
                    if name == 'socketcall':
                        subno = pintrace.socket.decode_call(pid)
                        subname = pinktrace.socket.name(subno)
                        if subname == 'connect':
                            addr = pinktrace.socket.decode_address(pid, 1)
                            break
                    elif name == 'connect':
                        addr = pinktrace.socket.decode_address(pid, 1)
                        break

            self.assert_(isinstance(addr, pinktrace.socket.Address), "%r" % addr)
            self.assertEqual(addr.family, socket.AF_INET)
            self.assertEqual(addr.ip, "127.0.0.1")
            self.assertEqual(addr.path, None)

            try: pinktrace.trace.kill(pid)
            except OSError: pass

if __name__ == '__main__':
    unittest.main()
